### DISCUSSION TITLE
Capturing StdOut from a process into a RichTextBox

### DISCUSSION THREAD

**Discussion by axtens** [unoplatform/uno#15283] [Q&A] [answered]
Out of the IDE we're building we're executing a process and capturing its stdout and stderr. Currently the stdout and stderr are gathered and then returned once the program has closed. I had wanted to display the stdout and stderr while they are being generated. However, referring to the RichEditBlock directly causes a `System.Runtime.InteropServices.COMException: 'The application called an interface that was marshalled for a different thread.` error.

I've now been alerted to `DispatcherQueue dispatcher = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();` which reputedly give me access to a thread outside those that will be created for .OutputDataReceived and .ErrorDataReceived. How would one use this dispatcher? Is this the thing to do?

**Comment by limeniye** [comment, marked as answer]
Try to use this sample:
https://github.com/unoplatform/uno/blob/69c31cc7c9c91cdaf1f5ae1fe0570f44190fd801/src/Uno.UI/Helpers/WinUI/DispatcherHelper.cs#L21

```C#
DispatcherQueue dispatcherQueue = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();
dispatcherQueue.TryEnqueue(() =>
{
    textBlock.Text = "Hello World!";
});
```

**Reply by axtens** [reply]
I ended up doing something very similar, declaring 
```
            DispatcherQueue dispatcher = Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread();
```
at the top and then having 
```
           proc.OutputDataReceived += async (object sender, DataReceivedEventArgs e) =>
           {
               if (quietude == 0)
               {
                   stdout.AppendLine(e.Data);
               }
               else
               {
                   await Task.Run(async () => { dispatcher.TryEnqueue(() => { AddOutput(e.Data!); }); });
               }
           };
           proc.ErrorDataReceived += async (object sender, DataReceivedEventArgs e) =>
           {
               if (quietude == 0)
               {
                   stderr.AppendLine(e.Data);
               }
               else
               {
                   await Task.Run(async () => { dispatcher.TryEnqueue(() => { AddError(e.Data!); }); });
               }
           };
```
where `dispatcher` is passed into the process manager and `AddOutput` and `AddError` wrap code to add paragraph runs to the RichEditBlock

**Reply by limeniye** [reply]
Your code hasn't sense.
Please, explain what have you tried to do and why do you need ``await Task.Run`` for that?
I could be wrong, but it looks like you're trying to make the code asynchronous, but it's actually just adding more load to your app. And as a result, there will be no asynchrony.

**Reply by axtens** [reply]
Well, it's what Visual Studio suggested and I'm ignorant. Now to try what you suggest.

**Reply by axtens** [reply]
So what you suggested works, but only the same as before. 
```
            proc.OutputDataReceived += (object sender, DataReceivedEventArgs e) =>
            {
                if (quietude == 0)
                {
                    stdout.AppendLine(e.Data);
                }
                else
                {
                    dispatcher.TryEnqueue(() =>
                    {
                        AddOutput(e.Data!);
                    });
                }
            };
```
Are you saying that asynchrony is impossible? That is, that the captured data cannot be seen to be filling up the RichTextBox as it is received?

**Reply by limeniye** [reply]
Writing to the RichTextBox can only be synchronous. This is how the dispatching of UI elements works. Therefore, this part ``await Task.Run(async () =>`` did not make any sense.
Your latest implementation looks more correct than the previous one.
